### PR TITLE
Enforce block version supermajority for BIP 66 (v3)

### PR DESCRIPTION
**Pull request by rnicoll**
Haven't completed testing on this yet, but wanted feedback on the general approach. Happy with the caching layer used to enable this? Any other comments before I invest more time on this?

I haven't had it apply BIP 34 partly because it's a bit late to be useful, and mostly because it would make Dogecoin support more of a problem for me later!

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/core/AbstractBlockChain.java [hunk 1]
```diff
@@ -466,13 +471,24 @@ private void connectBlock(final Block block, StoredBlock storedPrev, boolean exp
             }
             if (expensiveChecks && block.getTimeSeconds() <= getMedianTimestampOfRecentBlocks(head, blockStore))
                 throw new VerificationException("Block's timestamp is too early");
-            
+
+            // BIP 66: Enforce block version 3 once it's a supermajority of blocks
+            if (block.getVersion() == Block.BLOCK_VERSION_BIP34) {
+                final Integer count = versionTally.getCount(Block.BLOCK_VERSION_BIP66);
+                if (count != null
```
Comment (mikehearn) [inline core/src/main/java/org/bitcoinj/core/AbstractBlockChain.java]: Boxing? What's wrong with a good old fashioned -1?

File: core/src/main/java/org/bitcoinj/utils/VersionTally.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * Number of versions written into the tally. Until this matches the length
+     * of the version window, we do not have sufficient data to return values.
+     */
+    private int versionsStored = 0;
+
+    private final Object lock = new Object();
+
+    public      VersionTally(final NetworkParameters params) {
```
Comment (mikehearn) [inline core/src/main/java/org/bitcoinj/utils/VersionTally.java]: spacing

File: core/src/main/java/org/bitcoinj/utils/VersionTally.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    private int versionWriteHead = 0;
+
+    /**
+     * Number of versions written into the tally. Until this matches the length
+     * of the version window, we do not have sufficient data to return values.
+     */
+    private int versionsStored = 0;
+
+    private final Object lock = new Object();
```
Comment (mikehearn) [inline core/src/main/java/org/bitcoinj/utils/VersionTally.java]: What's wrong with just synchronizing on this?

File: core/src/main/java/org/bitcoinj/utils/VersionTally.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * @see org.bitcoinj.core.NetworkParameters#getMajorityWindow()
+ */
+public class VersionTally {
+    /**
+     * Current count for each version number. In theory could use an array for
+     * this, however means any out of sequence version numbers would require a
+     * much larger array.
+     */
+    private final Map<Long, Integer> counts = new HashMap<Long, Integer>();
```
Comment (mikehearn) [inline core/src/main/java/org/bitcoinj/utils/VersionTally.java]: The window isn't large. Counting each time is plenty fast enough. So why do we need this map?

File: core/src/main/java/org/bitcoinj/params/AbstractBitcoinNetParams.java [hunk 5]
```diff
@@ -40,6 +40,13 @@
      */
     public static final String BITCOIN_SCHEME = "bitcoin";
 
+    public static final int DEFAULT_MAJORITY_WINDOW = 1000;
+    public static final int DEFAULT_MAJORITY_REJECT_BLOCK_OUTDATED = 750;
+    public static final int DEFAULT_MAJORITY_ENFORCE_BLOCK_UPGRADE = 950;
+    public static final int TESTNET_MAJORITY_WINDOW = 100;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/params/AbstractBitcoinNetParams.java]: Can these testnet constants go into TestNetParams?

File: core/src/main/java/org/bitcoinj/params/AbstractBitcoinNetParams.java [hunk 6]
```diff
@@ -40,6 +40,13 @@
      */
     public static final String BITCOIN_SCHEME = "bitcoin";
 
+    public static final int DEFAULT_MAJORITY_WINDOW = 1000;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/params/AbstractBitcoinNetParams.java]: Those could be renamed to MAINNET_* and moved to MainNetParams?

**Comment by schildbach** [comment]
Does this code also work if you restore your wallet from a checkpoint, which ca be after the BIP66 rules came into effect?

**Comment by mikehearn** [comment]
@schildbach Good point. I suspect no. However checkpointing of wallets is not currently compatible with full verification.

It's probably easier/better to just hard code the fork time in the params, though. The voting thing is mostly useful for miners but nobody should be mining on top of bitcoinj.

File: core/src/main/java/org/bitcoinj/core/AbstractBlockChain.java (hunk 1 above)
Comment (rnicoll) [inline core/src/main/java/org/bitcoinj/core/AbstractBlockChain.java]: I like my return values to throw exceptions if you forget to check them. Want me to change it to an int?

File: core/src/main/java/org/bitcoinj/utils/VersionTally.java (hunk 3 above)
Comment (rnicoll) [inline core/src/main/java/org/bitcoinj/utils/VersionTally.java]: Good point, although in hindsight the class cannot be thread safe (the order of versions being added is important), so ripping the synchronization out anyway!

File: core/src/main/java/org/bitcoinj/utils/VersionTally.java (hunk 4 above)
Comment (rnicoll) [inline core/src/main/java/org/bitcoinj/utils/VersionTally.java]: Removed in next version

**Comment by rnicoll** [comment]
That should be most comments addressed in the two commits added, I'll squash at the end.

No, it won't do much if you're syncing from a checkpoint past the v3 switch over (or, it won't kick in until 1000 blocks after the checkpoint). It's really more meant so the code's in place and well tested for whatever the next soft fork is, especially given the v3 changeover did not go smoothly, and protecting against incorrectly mined blocks is a good idea in that context.

Could shelve this until actually needed, was mostly motivated while I remembered.

**Comment by schildbach** [comment]
Hmm, doesn't this code also fail if the peergroup stops and restarts later? I don't see any method of persisting the "tally". I wonder if it would be easier to just walk backwards the blocks in the block store and look at the version numbers. At least the block store is (usually) persisted.

**Comment by rnicoll** [comment]
Was thinking it would do a look back over the last 1000 blocks when creating the VersionTally (as said at the beginning, this isn't complete, I wanted feedback first), which doesn't help with checkpoints (the data just isn't there) but on a restart will work. Can just drop the cache if it's not worth the time saving, though.

**Comment by rnicoll** [comment]
Initialisation from block store added, testing completed, and ready to merge from my point of view. I can drop the cache, but am not sure looking back 1,000 blocks while validating what's currently most of the chain is a good idea (although yes, checkpoints mean this could be skipped entirely). Let me know what you think.

**Comment by schildbach** [comment]
I'm fine with merging this even with the cache, but I think two things would be nice:
* Unit tests
* A bit of explaination of the limits of this code (e.g. when starting from a checkpoint), and how this could be fixed in future.

**Comment by rnicoll** [comment]
Pushing work on unit tests for reference, but it's incomplete. I'll finish testing and add documentation on Thursday. Please do let me know if there's any problems with the unit tests so far, though.

**Comment by rnicoll** [comment]
Documentation added, unit tests done, rebased and squashed. Not sure what Travis is complaining about, again log doesn't show any problems until the last step when it returns 1 for some reason.

**Comment by schildbach** [comment]
I just hit the "rebuild" button on Travis and now it's green.

File: core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+import org.bitcoinj.store.BlockStore;
+import org.bitcoinj.store.BlockStoreException;
+import org.bitcoinj.store.MemoryBlockStore;
+import org.bitcoinj.testing.FakeTxBuilder;
+import org.junit.Test;
+import static org.junit.Assert.*;
+import org.junit.Before;
+
+/**
+ * Unit tests for the VersionTally class.
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java]: This comment feels redundant to me.

**Comment by schildbach** [comment]
To me this PR looks nice and ready to merge. And yes I'd prefer merging to just keeping it around for the next softfork, as it will likely bit-rot until then. @rnicoll What do you think?

**Comment by rnicoll** [comment]
@schildbach Merging sooner sounds good. I'll get the extra comment stripped out tonight.

**Comment by schildbach** [comment]
Don't worry, I just took the comment out and merged. Thanks again for your work!

File: core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    
+    @Test
+    public void testInitialize() throws BlockStoreException {
+        final BlockStore blockStore = new MemoryBlockStore(unitTestParams);
+        final BlockChain chain = new BlockChain(unitTestParams, blockStore);
+
+        // Build a historical chain of version 2 blocks
+        long timeSeconds = 1231006505;
+        StoredBlock chainHead = null;
+        for (int blockCount = 0; blockCount < unitTestParams.getMajorityWindow(); blockCount++) {
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java]: Late comment: If you assign unitTestParams = MainNetParams.get(), this loops runs forever. Looking at it more closely, I don't understand what you're trying to do and I wonder why it seems to work for UnitTestParams.

Comment (rnicoll) [inline core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java]: It's meant to test that the initialize() method actually fills VersionTally correctly. I suspect it doesn't work on main net due to the huge window main net uses, plus the difficulty (it essentially means it's trying to mine the first 1,000 blocks of the main network).

**Comment by TheBlueMatt** [comment]
Ehh, this is useless without also enforcing the (several) other softforks that bitcoinj still doesnt enforce, no?

**Comment by TheBlueMatt** [comment]
AFAIU bitcoinj still doesnt even enforce height-in-coinbase, so its full-node implementation is absolutely still a (pretty poor) SPV client.

File: core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java (hunk 8 above)
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java]: Hu, I didn't know we have mining code in bitcoinj… no we don't (-:

Seriously, even if you replace it by TestNet3Params (which has only a 100 window and a much lower difficulty, especially in the first 100 blocks), it runs into a seemingly endless loop. I suspect something fishy here.

**Comment by schildbach** [comment]
@TheBlueMatt We already agreed the code is of limited use as the switch has already happened, so it could be replaced by a hardcoded block height. We just want to have it around for the next soft fork.

File: core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java (hunk 8 above)
Comment (rnicoll) [inline core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java]: Could have sworn I saw code that guessed nonce values until one fitted, somewhere in createFakeBlock()...

Want to open an issue and assign it to me? Dealing with BIP 66 stuff, should get to it Thur/Fri

**Comment by rnicoll** [comment]
@TheBlueMatt Enforcing coinbase is on my todo-list (unless there's a reason it's impossible with SPV?)

File: core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java (hunk 8 above)
Comment (rnicoll) [inline core/src/test/java/org/bitcoinj/utils/VersionTallyTest.java]: https://github.com/bitcoinj/bitcoinj/blob/master/core/src/main/java/org/bitcoinj/core/Block.java#L635 - I'm not just crazy, see!

As said, open an issue so I've got a note, I'll confirm that's what's doing it, later this week

**Comment by TheBlueMatt** [comment]
@rnicoll Hmm? My point was updating the full-validation rules partially is of incredibly limited use (bitcoinj's full-validation is still only partial, implying it is effectively SPV security). I havent seen any push in a long time to update the full-validation code to be up-to-date. Really, it should just be removed.

**Comment by mikehearn** [comment]
I posted to the ML a while ago asking if anyone actually cares about the full validation code, or if it can be deleted. Nobody said they did. Agree that it should just be deleted. I think I pointed this out to Ross before.

**Comment by mikehearn** [comment]
That said, if Ross wants to step up and maintain the full validation code, that's fine by me.

Incomplete full validation is not completely useless. You wouldn't use it for mining, but there's a big spectrum of possibilities between full SPV and full validation. 

For instance a node that randomly validated half of all transactions in a block would double the cost for an attacking miner who was trying to fob you off with a block that spent non-existent/other peoples money, vs an attack on a pure SPV implementation where they are guaranteed to eventually succeed. That cost increase might make the attack unprofitable, especially in the presence of other security mechanisms (like if you get one shot before your expensive customer account is banned).

**Comment by TheBlueMatt** [comment]
Sure, running incomplete full validation for the purposes of lowering CPU load is a reasonable tradeoff. Soft forks, however, are a pretty important part of partial SPV validation, however, as seen by the SPV mining issues recently. So, I would say if we cant even get soft forks from a year or more implemented, it may still be worse to have the feature than to not have it.

**Comment by TheBlueMatt** [comment]
Whats worse, the script implementation used is still the built-in Java one, which absolutely should be removed now that there is an easy alternative.

**Comment by rnicoll** [comment]
Coinbase is of course impossible to check because SPV means the client won't necessarily have the coinbase transaction. Would it be good enough if it checks it if it has it?

Either way, not clear on how having this could be worse?

**Comment by TheBlueMatt** [comment]
I was suggesting if you were doing a partial-validation full node (ie partially-spv) one of the most important things to do is check for soft forks you want to enforce, not that they should be enforced by SPV clients (though, they should as much as possible, but, yea, getting the coinbase in every block would suck).

**Comment by mikehearn** [comment]
Yes, you may be right. I haven't really decided whether to delete the full validation code or not. But if nobody steps up to implement height-in-coinbase soon then I'll remove it.

@rnicoll there's no point in SPV wallets trying to verify coinbase transactions.

**Comment by rnicoll** [comment]
@mikehearn Happy to commit to adding in height-in-coinbase for full validation, but likely to be towards the end of the month.

**Comment by mikehearn** [comment]
There's no rush. If you're willing to maintain the code then alright, we can keep it. It'll still have some health warnings attached because consensus bugs are very likely still lurking, but if people find it useful and are willing to take the risk anyway then OK.

**Pull request closed without merging**